aip/client-libraries/4223.md

```diff
@@ -0,0 +1,39 @@
+---
+id: 4223
+state: draft # TODO(noahdietz): Update pre-merge
+created: 2022-05-11 # TODO(noahdietz): Update pre-merge
+---
+
+# Client-side payload validation
+
+Services document the behavior of specific request fields using the
+`google.api.field_behavior` annotation defined in [AIP-203][]. This annotation
+conveys how a service interprets the field, including how a field may be
+validated (or not) by the service.
+
+## Guidance
+
+Client libraries **must not** implement client-side payload validation based
+on the `google.api.field_behavior` annotation, except to prevent
+[local failures](4223.md#local-failures).
+
+The `google.api.field_behavior` annotation is primarily a machine-readable
+form of _documentation_, and **not** a configuration for payload validation. The
+annotation conveys how the service _already_ handles the field, and alludes to
+a service's own payload validation or expectations.
+
+Admittedly, depending on services to implement payload validation means clients
+will sometimes make network calls that are bound to result in errors. Furthermore, users
+then depend on services to provide actionable errors, when the client could have
+provided a language-idiomatic, localized error. However, robust server-side
+payload validation means that _all types_ of client (first party, third party,
+`curl`, etc.) will benefit from the **same** validation, and that there is no
+lag between service and clients when the field behavior changes.
+
+### Local failures
```

Comment:
Might be good to define "local failures". I'm not sure what this means, nor can I find this term referenced in any of the cross-referenced AIPs.

Reply:
Kind of similar to my response to Daniel's comment, it really is a language-dependent thing. Basically, validation of the request prior to the network call is acceptable to the extent that it prevents Null Pointer Exceptions, or a `panic`, or whatever is the idiomatic "crash" scenario in each language. I'd rather not get too specific because I don't want to enumerate something for every language, especially when something might be an error scenario in one language but not in another.

> nor can I find this term referenced in any of the cross-referenced AIPs.

It is a client code specific thing, not going to be in API Design AIPs.

Reply:
Maybe change it to be,

> Client libraries **may** implement client-side payload validation based on the
`google.api.field_behavior` annotation only to the extent that it prevents
local failures, such as crashes within the client library code itself.

Reply:
Thanks, that sounds like a great addition. Will do